src/task_manager.cc

```diff
@@ -43,13 +43,17 @@ namespace detail {
 	}
 
 	void task_manager::notify_horizon_executed(task_id tid) {
+#ifndef NDEBUG
+		std::lock_guard lock{task_mutex};
+		assert(task_map.count(tid) != 0);
 		assert(task_map.at(tid)->get_type() == task_type::HORIZON);
 		assert(executed_horizons.empty() || executed_horizons.back() != tid);
-		executed_horizons.push(tid);
+#endif
 
+		executed_horizons.push(tid); // no locking needed - see definition
 		if(executed_horizons.size() >= horizon_deletion_lag) {
 			// actual cleanup happens on new task creation
-			horizon_task_id_for_deletion = executed_horizons.front();
+			horizon_task_id_for_deletion.store(executed_horizons.front()); // atomic
```

Comment:
I don't understand this change -- the assignment operator on `std::atomic`s is atomic.

Reply:
That's clarification only. `notify_horizon_executed` doesn't take a lock (in release builds) but still accesses data members, so I figure it's good to know why that is safe.

Reply:
I feel like this is a bit over-clarified with using `.store` and `// atomic` if the type is already `atomic`, but it's not big deal.